src/dwarf_parser.cpp

```diff
@@ -1,250 +1,216 @@
+        case lldb::eBasicTypeChar32:
+        case lldb::eBasicTypeShort:
+        case lldb::eBasicTypeInt:
+        case lldb::eBasicTypeInt128:
+        case lldb::eBasicTypeLong:
+        case lldb::eBasicTypeLongLong:
           return CreateInt(bit_size);
+        case lldb::eBasicTypeUnsignedChar:
+        case lldb::eBasicTypeUnsignedWChar:
+        case lldb::eBasicTypeUnsignedShort:
+        case lldb::eBasicTypeUnsignedInt:
+        case lldb::eBasicTypeUnsignedInt128:
+        case lldb::eBasicTypeUnsignedLong:
+        case lldb::eBasicTypeUnsignedLongLong:
+          return CreateUInt(bit_size);
         default:
           return CreateNone();
       }
+      break;
     }
-    case DW_TAG_enumeration_type:
+    case lldb::eTypeClassEnumeration:
       return CreateUInt(bit_size);
-    case DW_TAG_pointer_type: {
-      if (dwarf_hasattr(&type, DW_AT_type)) {
-        Dwarf_Die inner_type = type_of(type);
+    case lldb::eTypeClassPointer:
+    case lldb::eTypeClassReference: {
+      if (auto inner_type = type.GetPointeeType()) {
         return CreatePointer(get_stype(inner_type, false));
       }
       // void *
       return CreatePointer(CreateNone());
     }
-    case DW_TAG_structure_type:
-    case DW_TAG_union_type: {
-      std::string name = dwarf_diename(&type_die);
-      name = (tag == DW_TAG_structure_type ? "struct " : "union ") + name;
+    case lldb::eTypeClassClass:
+    case lldb::eTypeClassStruct:
+    case lldb::eTypeClassUnion: {
+      std::string name = type.GetDisplayTypeName() ?: "<unknown type>";
+      if (type.GetTypeClass() == lldb::eTypeClassClass) {
+        name = "class " + name;
+      } else if (type.GetTypeClass() == lldb::eTypeClassStruct) {
+        name = "struct " + name;
+      } else if (type.GetTypeClass() == lldb::eTypeClassUnion) {
+        name = "union " + name;
+      } else {
+        // unreachable
+      }
       auto result = CreateRecord(
           name, bpftrace_->structs.LookupOrAdd(name, bit_size / 8));
       if (resolve_structs)
         resolve_fields(result);
       return result;
     }
-    case DW_TAG_array_type: {
-      Dwarf_Die inner_type_die = type_of(type_die);
-      Dwarf_Word inner_enc = get_type_encoding(inner_type_die);
-      SizedType inner_type = get_stype(inner_type_die);
-      SizedType result;
-      for (auto &d :
-           get_all_children_with_tag(&type_die, DW_TAG_subrange_type)) {
-        ssize_t size = get_array_size(d);
-        if (dwarf_tag(&inner_type_die) == DW_TAG_base_type &&
-            (inner_enc == DW_ATE_signed_char ||
-             inner_enc == DW_ATE_unsigned_char))
-          result = CreateString(size);
-        else
-          result = CreateArray(size, inner_type);
-        inner_type = result;
+    case lldb::eTypeClassArray: {
+      auto inner_type = type.GetArrayElementType();
+      auto inner_stype = get_stype(inner_type);
+      auto size = type.GetByteSize() / inner_type.GetByteSize();
```

Comment:
What about multi-dimensional arrays? IIRC, DWARF doesn't represent them as "arrays of arrays" but as "arrays with multiple dimensions". Will this handle them?

Reply:
I've added a test for arrays, see "tests/field_analyser.cpp". It works as I would expect 🙂

Reply:
Are multi-dimensional arrays getting flattened down to 1-D arrays? It'd be nice if users could access them in the same way as in C/C++ source code, i.e. `$foo->d[0][2]`

Reply:
Printing the SizedType of the test array does give me int32[4][4], so the SizedType properly represents the array.

Reply:
Ah cool! I think I'd misread the new test case you added for it.

It would be good to check the full details of the array in the test case too, to ensure it stays working this way in the future, e.g. `GetNumElements()`, `GetElementTy()` for both the outer and inner arrays